src/Workspaces/Core/Portable/TemporaryStorage/TemporaryStorageServiceFactory.cs

```diff
@@ -101,14 +101,22 @@ public ITemporaryTextStorageInternal CreateTemporaryTextStorage()
         string storageName, long offset, long size, SourceHashAlgorithm checksumAlgorithm, Encoding? encoding, ImmutableArray<byte> contentHash)
         => new(this, storageName, offset, size, checksumAlgorithm, encoding, contentHash);
 
-    ITemporaryStreamStorageInternal ITemporaryStorageServiceInternal.CreateTemporaryStreamStorage()
-        => CreateTemporaryStreamStorage();
+    public TemporaryStorageHandle WriteToTemporaryStorage(Stream stream, CancellationToken cancellationToken)
+    {
+        var storage = new TemporaryStreamStorage(this);
+        storage.WriteStream(stream, cancellationToken);
+        var identifier = new TemporaryStorageIdentifier(storage.Name, storage.Offset, storage.Size);
+        return new(storage, identifier);
+    }
 
-    internal TemporaryStreamStorage CreateTemporaryStreamStorage()
-        => new(this);
+    Stream ITemporaryStorageServiceInternal.ReadFromTemporaryStorageService(TemporaryStorageIdentifier storageIdentifier, CancellationToken cancellationToken)
+        => ReadFromTemporaryStorageService(storageIdentifier, cancellationToken);
 
-    public TemporaryStreamStorage AttachTemporaryStreamStorage(string storageName, long offset, long size)
-        => new(this, storageName, offset, size);
```

Comment:
this was the worst part of the old system.  You'd first ask to create a temp storage.  Then ask to attach it to the data, getting another temp storage.  But then that temp storage was only readable, while the original one was only writable.  it was a confusing mess of an api that had all sorts of rules on how to use it properly.